File: packages/engine/src/protocol/domains/system.ts
import { Schema } from "effect";
import {
  NonEmptyString,
  PositiveInt,
  optionalWith,
} from "@guerillaglass/engine/protocol/shared/helpers";

/** Result payload for `system.ping`. */
export const pingResultSchema = Schema.Struct({
  app: NonEmptyString,
  engineVersion: NonEmptyString,
  protocolVersion: NonEmptyString,
  platform: NonEmptyString,
});

const capabilitiesAgentSchema = Schema.Struct({
  preflight: Schema.Boolean,
  run: Schema.Boolean,
  status: Schema.Boolean,
  apply: Schema.Boolean,
  localOnly: optionalWith(Schema.Boolean, { default: () => true }),
  runtimeBudgetMinutes: optionalWith(PositiveInt, { default: () => 10 }),
});

/** Result payload for `engine.capabilities`. */
export const capabilitiesResultSchema = Schema.Struct({
  protocolVersion: NonEmptyString,
  platform: NonEmptyString,
  phase: Schema.Literals(["stub", "foundation", "native"]),
  capture: Schema.Struct({
    display: Schema.Boolean,
    window: Schema.Boolean,
    systemAudio: Schema.Boolean,
    microphone: Schema.Boolean,
  }),
  recording: Schema.Struct({
    inputTracking: Schema.Boolean,
  }),
  export: Schema.Struct({
    presets: Schema.Boolean,
    cutPlan: optionalWith(Schema.Boolean, { default: () => false }),
  }),
  project: Schema.Struct({
    openSave: Schema.Boolean,
  }),
  agent: optionalWith(capabilitiesAgentSchema, {
    default: () => ({
      preflight: false,
      run: false,
      status: false,
      apply: false,
      localOnly: true,
      runtimeBudgetMinutes: 10,
    }),
  }),
});

/** Type alias for PingResult. */
export type PingResult = Schema.Schema.Type<typeof pingResultSchema>;
/** Type alias for CapabilitiesResult. */
export type CapabilitiesResult = Schema.Schema.Type<typeof capabilitiesResultSchema>;

File: engines/native-foundation/src/system.rs
use crate::wire::{success, EngineCallId, EngineResponse, PROTOCOL_VERSION};
use crate::{ENGINE_PHASE, ENGINE_VERSION};
use serde_json::json;

pub(crate) fn ping(id: &EngineCallId, platform: &str) -> EngineResponse {
    success(
        id,
        json!({
            "app": "guerillaglass",
            "engineVersion": ENGINE_VERSION,
            "protocolVersion": PROTOCOL_VERSION,
            "platform": platform,
        }),
    )
}

pub(crate) fn capabilities(id: &EngineCallId, platform: &str) -> EngineResponse {
    success(
        id,
        json!({
            "protocolVersion": PROTOCOL_VERSION,
            "platform": platform,
            "phase": ENGINE_PHASE,
            "capture": {
                "display": true,
                "window": true,
                "systemAudio": true,
                "microphone": true,
            },
            "recording": {
                "inputTracking": true,
            },
            "export": {
                "presets": true,
                "cutPlan": true,
                "backgroundFraming": false,
            },
            "project": {
                "openSave": true,
            },
            "agent": {
                "preflight": true,
                "run": true,
                "status": true,
                "apply": true,
                "localOnly": true,
                "runtimeBudgetMinutes": 10,
            }
        }),
    )
}
